src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Web/Debugger/WebLaunchSettingsProvider.cs

```diff
@@ -0,0 +1,301 @@
+    {
+
+        private static class ProjectProperties
+        {
+            public const string UseIISExpress = nameof(UseIISExpress);
+            public const string Use64BitIISExpress = nameof(Use64BitIISExpress);
+            public const string IISExpressWindowsAuthentication = nameof(IISExpressWindowsAuthentication);
+            public const string IISExpressAnonymousAuthentication = nameof(IISExpressAnonymousAuthentication);
+            public const string IISExpressSSLPort = nameof(IISExpressSSLPort);
+            public const string IISExpressUseClassicPipelineMode = nameof(IISExpressUseClassicPipelineMode);
+            public const string UseGlobalApplicationHostFile = nameof(UseGlobalApplicationHostFile);
+        }
+
+        private const string WAPProjectGuidString = "{" + ProjectType.LegacyWebApplicationProject + "}";
+        private readonly IUnconfiguredProjectCommonServices _projectServices;
+        private readonly IProjectExtensionDataService _projectExtensionDataService;
+        private FlavorServerSettings? _flavorServerSettings;
+
+        private bool _useIISExpress;
+        private bool _use64BitIISExpress;
+        private int _issExpressSSLPort;
+        private bool _useWindowsAuth;
+        private bool _useAnonymousAuth;
+        private bool _useClassicPipelineMode;
+        private bool _useGlobalApplicationHostFile;
+
+        [ImportingConstructor]
+        public WebLaunchSettingsProvider(
+            ILaunchSettingsProvider2 launchSettingsProvider,
+            IUnconfiguredProjectCommonServices projectServices,
+            IProjectExtensionDataService projectExtensionDataService,
+            IProjectThreadingService threadingService)
+             : base(threadingService.JoinableTaskContext)
+        {
+            _projectServices = projectServices;
+            _projectExtensionDataService = projectExtensionDataService;
+        }
+
+        public async Task<WebLaunchSettings> GetLaunchSettingsAsync()
+        {
+            await InitializeAsync();
+
+            List<string> urls = new List<string>();
+            ServerType serverType = ServerType.IISExpress;
+
+            if (_flavorServerSettings == null)
+            {
+                return new WebLaunchSettings(serverType, urls, _useWindowsAuth, _useAnonymousAuth, _useClassicPipelineMode, _use64BitIISExpress, _useGlobalApplicationHostFile, false, null);
+            }
+
+            if (_flavorServerSettings.UseCustomServer)
+            {
+                if (_flavorServerSettings.CustomServerUrl != null)
+                {
+                    urls.Add(_flavorServerSettings.CustomServerUrl);
+                }
+
+                serverType = ServerType.Custom;
+            }
+            else if (_flavorServerSettings.UseIIS)
+            {
+                if (_useIISExpress)
+                {
+                    if (_flavorServerSettings.IISUrl != null)
```

Comment:
>IISUrl != null [](start = 46, length = 14)

does this need a string emptiness check?

Reply:
I added one,. Do you know how to avoid the "CS8604: Possible Null reference argument" I get on the urls.Add line? The analyzer is not smart enough to recognize that string.IsNullOrEmpty is a null check. I realize I can pragma disable it but that is pretty ugly.
                    if (!string.IsNullOrEmpty(_flavorServerSettings.IISUrl))
                    {
                        urls.Add(_flavorServerSettings.IISUrl);

 Or I guess I can do the checks myself: ie 
if  (_flavorServerSettings.IISUrl) != null && _flavorServerSettings.IISUrl.Length != 0)

Reply:
I just did the check myself